src/_pytest/nodes.py

```diff
@@ -106,6 +111,34 @@ def __init__(
             if self.name != "()":
                 self._nodeid += "::" + self.name
 
+        if not type(self).__in_named_ctor:
+            warnings.warn(
+                NODE_USE_FROM_PARENT.format(name=type(self).__name__),
+                stacklevel=3,
+            )
+
+    @classmethod
+    @contextmanager
+    def __ctor_entered(cls):
+        cls.__in_named_ctor = True
+        yield
+        cls.__in_named_ctor = False
+
+    @staticmethod
+    def constructor(ctor):
```

Comment:
This mechanism is quite complex and is not reentrant.

A more direct option would be to add to each constructor a "hidden" kwonly parameter `__called_by_named_ctor=False`. If it's false, show warning, if it's true, don't. Then the named ctors would set it. On the next breaking release, remove it and use a metaclass like [this](https://github.com/python-trio/trio/blob/d36c2a5e0c3c1a4d93cc2fd35023f9519fd7205b/trio/_util.py#L301-L323).

Reply:
good link/point, that mechanism is better than mine, i will in cooperate it in a bit